planning/behavior_path_planner/src/scene_module/avoidance/avoidance_utils.cpp

```diff
@@ -264,6 +263,305 @@ tier4_autoware_utils::Polygon2d expandPolygon(
   return expanded_polygon;
 }
 
+boost::optional<geometry_msgs::msg::Point> intersect(
+  const geometry_msgs::msg::Point & p1, const geometry_msgs::msg::Point & p2,
+  const geometry_msgs::msg::Point & p3, const geometry_msgs::msg::Point & p4)
+{
+  // calculate intersection point
+  const double det = (p1.x - p2.x) * (p4.y - p3.y) - (p4.x - p3.x) * (p1.y - p2.y);
+  if (det == 0.0) {
+    return {};
+  }
+
+  const double t = ((p4.y - p3.y) * (p4.x - p2.x) + (p3.x - p4.x) * (p4.y - p2.y)) / det;
+  const double s = ((p2.y - p1.y) * (p4.x - p2.x) + (p1.x - p2.x) * (p4.y - p2.y)) / det;
+  if (t < 0 || 1 < t || s < 0 || 1 < s) {
+    return {};
+  }
+
+  geometry_msgs::msg::Point intersect_point;
+  intersect_point.x = t * p1.x + (1.0 - t) * p2.x;
+  intersect_point.y = t * p1.y + (1.0 - t) * p2.y;
+  intersect_point.z = t * p1.z + (1.0 - t) * p2.z;
+  return intersect_point;
+}
+
+boost::optional<std::pair<size_t, geometry_msgs::msg::Point>> intersectBound(
+  const geometry_msgs::msg::Point & p1, const geometry_msgs::msg::Point & p2,
+  const std::vector<geometry_msgs::msg::Point> & bound, const size_t seg_idx1,
+  const size_t seg_idx2)
+{
+  const size_t start_idx =
+    static_cast<size_t>(std::max(0, static_cast<int>(std::min(seg_idx1, seg_idx2)) - 5));
+  const size_t end_idx = static_cast<size_t>(std::min(
+    static_cast<int>(bound.size()) - 1, static_cast<int>(std::max(seg_idx1, seg_idx2)) + 1 + 5));
+  for (int i = start_idx; i < static_cast<int>(end_idx); ++i) {
+    const auto intersect_point = intersect(p1, p2, bound.at(i), bound.at(i + 1));
+    if (intersect_point) {
+      std::pair<size_t, geometry_msgs::msg::Point> result;
+      result.first = static_cast<size_t>(i);
+      result.second = intersect_point.get();
+      return result;
+    }
+  }
+  return boost::none;
+}
+
+PolygonPoint transformBoundFrenetCoordinate(
+  const std::vector<geometry_msgs::msg::Point> & points, const geometry_msgs::msg::Point & point)
+{
+  const size_t seg_idx = motion_utils::findNearestSegmentIndex(points, point);
+  const double lon_dist_to_segment =
+    motion_utils::calcLongitudinalOffsetToSegment(points, seg_idx, point);
+  const double lat_dist = motion_utils::calcLateralOffset(points, point, seg_idx);
+  return PolygonPoint{point, seg_idx, lon_dist_to_segment, lat_dist};
+}
+
+std::vector<PolygonPoint> generatePolygonInsideBounds(
+  const std::vector<Point> & bound, const std::vector<Point> & edge_points,
+  const bool is_object_right)
+{
+  std::vector<PolygonPoint> full_polygon;
+  for (size_t i = 0; i < edge_points.size(); ++i) {
+    const auto polygon_point = transformBoundFrenetCoordinate(bound, edge_points.at(i));
+    full_polygon.push_back(polygon_point);
+  }
```

Comment:
I think `full_polygon` can be `reserve`.
Is there a reason for not using range-based for loop here?

Reply:
I agree to use range-based for loop.
Fixed. https://github.com/autowarefoundation/autoware.universe/pull/2880/commits/2f6b217d9f8a572831f5e8b36134f1ce9f4a6be3